### ISSUE TITLE
Weird siprintf behaviour

### ISSUE THREAD

**Issue by ReservedField** [ReservedField/evic-sdk#16]
I don't know where this happened, but `Display_PutText` on constant strings seems to be broken. Actually, it's broken in a very weird way. Test case:

```
#include <stdio.h>
#include <M451Series.h>
#include <Display.h>
#include <Font.h>

int main() {
    char buf[100];

    // Without this line it doesn't work
    siprintf(buf, "whatever");

    Display_PutText(8, 56, "Hello,\nWorld.", FONT_DEJAVU_8PT);
    Display_Update();
}
```

It seems to crash when removing the `siprintf` line. Which writes on a buffer that is never used. I'm investigating this, but still have no idea what is happening.

**Comment by ReservedField** [comment]
Further investigation shows that no code runs unless siprintf is called somewhere in the program. It can be called after the code, but it needs to be called. Simply ensuring it is imported does not work. This is clearly a compile-time issue, but I'm a bit lost as to what causes it...

**Comment by zygisber** [comment]
Full moon :-) At least in my country :-) Dirty work arround will be include
it was in init functions in startup, maybe... :-D

**Comment by ReservedField** [comment]
Source of regression is commit 01137a0.

**Comment by chaosman123** [comment]
Are you sure it was caused by that specific commit?
Tried to remove the different parts, but still did not work.
Ok, did not roll back all parts, but the major ones...

**Comment by ReservedField** [comment]
I tried rolling back and forth and that revision indeed caused the issue. You can't "partially" roll it back, it's a major change in the build process.

**Comment by chaosman123** [comment]
Found some time to dig a bit futher. Problem solved.
The problem was not only caused by the given commitment - as I already said.
It went hand in hand with the commitment before. Attached you find a working linker.ld.

```
/* Memory regions */
MEMORY {
        ROM (rx)  : ORIGIN = 0x00000000, LENGTH = 128K
        RAM (rwx) : ORIGIN = 0x20000000, LENGTH =  32K
}

/* Library configurations */
INPUT(libevicsdk_crt0.o)
GROUP(
        armv7e-m/libgcc.a
        armv7e-m/libc.a
        armv7e-m/libm.a
        libevicsdk.a
)

ENTRY(Reset_Handler)

SECTIONS {
        .text : {
                KEEP(*(.isr_vector))
                *(.text*)

                KEEP(*(.init))

                . = ALIGN(4);
                __preinit_array_start = .;
                KEEP(*(.preinit_array))
                __preinit_array_end = .;
                __init_array_start = .;
                KEEP(*(.init_array))
                __init_array_end = .;
                KEEP(*(.fini))
                . = ALIGN(4);
                __fini_array_start = .;
                KEEP(*(.fini_array))
                __fini_array_end = .;

                /* .ctors */
                *crtbegin.o(.ctors)
                *crtbegin?.o(.ctors)
                *(EXCLUDE_FILE(*crtend?.o *crtend.o) .ctors)
                *(SORT(.ctors.*))
                *(.ctors)

                /* .dtors */
                *crtbegin.o(.dtors)
                *crtbegin?.o(.dtors)
                *(EXCLUDE_FILE(*crtend?.o *crtend.o) .dtors)
                *(SORT(.dtors.*))
                *(.dtors)

                *(.rodata*)

                KEEP(*(.eh_frame*))
        } > ROM

        .ARM.extab : {
                *(.ARM.extab* .gnu.linkonce.armextab.*)
        } > ROM

        .ARM.exidx : {
                *(.ARM.exidx* .gnu.linkonce.armexidx.*)
        } > ROM

        Data_Start_ROM = .;

        .data : AT (Data_Start_ROM) {
                Data_Start_RAM = .;
                *(vtable)
                *(.data*)

                . = ALIGN(4);
                KEEP(*(.preinit_array))

                . = ALIGN(4);
                KEEP(*(SORT(.init_array.*)))
                KEEP(*(.init_array))

                . = ALIGN(4);
                KEEP(*(SORT(.fini_array.*)))
                KEEP(*(.fini_array))

                KEEP(*(.evicsdk_tag))
                KEEP(*(.verify))

                . = ALIGN(4);
                Data_End_RAM = .;
        } > RAM

        Data_Size = Data_End_RAM - Data_Start_RAM;

        .bss : {
                . = ALIGN(4);
                BSS_Start = .;
                *(.bss*)
                *(COMMON)
                . = ALIGN(4);
                BSS_End = .;
        } > RAM

        BSS_Size = BSS_End - BSS_Start;

        .stack : {
                /* Stack must be 8-byte aligned for double local vars */
                . = ALIGN(8);
                *(.stack)
        } > RAM

        /* Stack top to end of RAM */
        Stack_Top = ORIGIN(RAM) + LENGTH(RAM);
        Stack_Limit = Stack_Top - SIZEOF(.stack);

        /* Fill all unused RAM with heap */
        Heap_Start = BSS_End;
        Heap_Limit = Stack_Limit;

        /* Check for RAM overflow */
        ASSERT(Stack_Limit > Heap_Start, "RAM overflowed with stack")
}
```

**Comment by ReservedField** [comment]
@chaosman123 I'll be able to test in a few hours. The differences I see are that you moved `.fini` from after `.init` to before `.fini_array`. You also moved `.evicsdk_tag` and `.verify` to the `.data` section.
(I have no idea why I also put preinit/init/fini arrays in `.data` - those should be in ROM).
The only issue I see with this is that `.evicsdk_tag` and `.verify` are moved to `.data` - this means they'll be copied to RAM on startup, which is not something we want (it won't cause problems, but it will waste some RAM).
I'll do more test as soon as I get home.

**Comment by chaosman123** [comment]
Not sure what your idea behind that tag is... never in the past used something like that.
I just added it back in - as far as I remeber, where it was.
The tag was not the problem. That's just what I can tell you.
Currently don't have access to the source, so cannot create a diff here.

**Comment by ReservedField** [comment]
The tag is just for debugging, each binary is marked with the SDK revision. Most compilers do it.
I'll try to play with it some more.

**Comment by chaosman123** [comment]
BTW, how do we continue work? Shall I send patches? Shall we work on different trees?
Second Cuboid and Intronix logic analyzer are on the way to me ;)
Sort of a forum would be good for discussions, plus a Wiki for Dokumentation (PCB screenshots, datasheets, APROM findings, etc.).
Never used Github before, but does not look like the right place for those kind of things...

**Comment by Ban3** [comment]
> Sort of a forum would be good for discussions

There's an IRC channel, ##evic-modding on Freenode, for realtime chatting. We could create a subreddit if there's enough interest for a forum style dicussions.

> plus a Wiki for Dokumentation 

Github provides this.

**Comment by ReservedField** [comment]
@chaosman123: Sorry, I've been really busy for a while now. If you have patches send them my way, or make a branch, patch it and open a pull request.

**Comment by ReservedField** [comment]
The tag/verify sections caused the `.data` section to become misaligned. The startup code assumes a 4-byte alignment for the ROM to RAM copy, so it caused an infinite loop and crash. Fixed in 6eaebfc.
